src/NodaTime.Test/OffsetDateTest.cs

```diff
@@ -0,0 +1,138 @@
+        {
+            var date = new LocalDate(2012, 1, 2);
+            var offset = Offset.FromHours(5);
+
+            var offsetDate = new OffsetDate(date, offset);
+            Assert.AreEqual(offset, offsetDate.Offset);
+            Assert.AreEqual(date, offsetDate.Date);
+        }
+
+        [Test]
+        public void Equality()
+        {
+            LocalDate date1 = new LocalDate(2012, 10, 6);
+            LocalDate date2 = new LocalDate(2012, 9, 5);
+            Offset offset1 = Offset.FromHours(1);
+            Offset offset2 = Offset.FromHours(2);
+
+            OffsetDate equal1 = new OffsetDate(date1, offset1);
+            OffsetDate equal2 = new OffsetDate(date1, offset1);
+            OffsetDate unequalByOffset = new OffsetDate(date1, offset2);
+            OffsetDate unequalByLocal = new OffsetDate(date2, offset1);
+
+            TestHelper.TestEqualsStruct(equal1, equal2, unequalByOffset);
+            TestHelper.TestEqualsStruct(equal1, equal2, unequalByLocal);
+
+            TestHelper.TestOperatorEquality(equal1, equal2, unequalByOffset);
+            TestHelper.TestOperatorEquality(equal1, equal2, unequalByLocal);
+        }
+
+        [Test]
+        public void At()
+        {
+            var date = new LocalDate(2012, 6, 19, CalendarSystem.Julian);
+            var offset = Offset.FromHours(5);
+            var time = new LocalTime(14, 15, 12).PlusNanoseconds(123456789);
+
+            Assert.AreEqual(new OffsetDate(date, offset).At(time), date.At(time).WithOffset(offset));
+        }
+
+        [Test]
+        public void WithOffset()
+        {
+            var date = new LocalDate(2012, 6, 19);
+            var initial = new OffsetDate(date, Offset.FromHours(2));
+            var actual = initial.WithOffset(Offset.FromHours(5));
+            var expected = new OffsetDate(date, Offset.FromHours(5));
+            Assert.AreEqual(expected, actual);
+        }
+
+        [Test]
+        public void WithDate()
+        {
+            var offset = Offset.FromHours(2);
+            var initial = new OffsetDate(new LocalDate(2012, 6, 19), offset);
+            var newDate = new LocalDate(2015, 1, 2);
+            var actual = initial.WithDate(newDate);
+            var expected = new OffsetDate(newDate, offset);
+            Assert.AreEqual(expected, actual);
+        }
+
+        [Test]
+        public void WithCalendar()
+        {
+            var julianDate = new LocalDate(2012, 6, 19, CalendarSystem.Julian);
+            var isoDate = julianDate.WithCalendar(CalendarSystem.Iso);
+            var offset = Offset.FromHours(5);
+            var actual = new OffsetDate(julianDate, offset).WithCalendar(CalendarSystem.Iso);
+            var expected = new OffsetDate(isoDate, offset);
+            Assert.AreEqual(expected, actual);
+        }
+
+        [Test]
+        public void WithAdjuster()
+        {
+            var initial = new OffsetDate(new LocalDate(2016, 6, 19), Offset.FromHours(5));
```

Comment:
Well, the new date would *start* before the start of the UTC month, but I think it would be a slightly odd test. (In order to demonstrate that it's before the start of the month, I'd need to convert both into OffsetDateTime values or Instants.)